### PR TITLE
ui: pretty-print statement fingerprints

### PR DESCRIPTION
**Pull request by couchand**
Thanks to @mjibson in #26943 there is now a pretty-printer built in to Cockroach.  This leverages that to nicely format the statement fingerprints for the Statements pages.

details page:
screen shot 2018-07-03 at 11 37 20 am

tooltip:
screen shot 2018-07-03 at 12 18 51 pm

### CODE REVIEW

**Comment by couchand** [comment]
I'm not sure how I feel about running every statement fingerprint through the parser for this endpoint.  Also the changes from the second commit feel really gross.

But it's fun and neat!

File: pkg/ui/src/util/sql/summarize.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ export interface StatementSummary {
 
 const keywords: { [key: string]: RegExp } = {
   update: /^update\s+(\S+)/i,
-  select: /^select.+from\s+(\S+)/i,
+  select: /^select(?:fro[^m]|fr[^o]|f[^r]|[^f])+from\s+(\S+)/i,
```
Comment (vilterp) [inline pkg/ui/src/util/sql/summarize.ts]: what's the story behind this? stretching my regex-foo.

**Review by vilterp — COMMENTED** [review]
Looks good! Wouldn't expect perf to be too much of a problem, but maybe we should add a little microbenchmark to see how fast it is.

**Comment by vilterp** [comment]
Talked IRL with @mjibson: as a result of some optimization work with @knz (can't find the PR), the pretty printer should now be fast enough to run here, even twice per fingerprint. It runs even the biggest queries we're likely to see ("pages long") in tens of ms and megs of memory.

**Comment by vilterp** [comment]
Hm, tried it out on a single-node cluster on my laptop running TPCC and it may still be a little too slow for us to do pretty printing server-side: the statements endpoint took about 1s to return while the loadgen is running, and 0.8s while loadgen not running.

Screenshot of the CPU profile:
image

**Comment by knz** [comment]
yes I disagree with the assessment that it is reasonable to call this pretty-printer for every entry in the stmt stats.

Instead we should have a RPC endpoint to pretty-print stuff, and the UI code should call that endpoint lazily to render things on demand.

**Comment by mjibson** [comment]
Hmm I thought it was faster. What query did you try?

**Comment by knz** [comment]
@mjibson he's not printing just 1, he's printing all of them, upfront:

https://github.com/vilterp/cockroach/commit/110ce877df0d2816ed0af26ff7bd49b34dd8c7dc#diff-2cd4249d9132b7be0cd20309f3160490R94

That's unreasonable.

File: pkg/server/statements.go [hunk 2]
```diff
@@ -89,13 +91,21 @@ func (s *statusServer) StatementsLocal(ctx context.Context) (*serverpb.Statement
 	}
 
 	for i, stmt := range stmtStats {
+		stmts, err := parser.Parse(stmt.Key.Query)
+		if err != nil {
+			return nil, err
+		}
+		pretty := tree.PrettyWithOpts(stmts[0], 80, false, 4)
+		prettyNarrow := tree.PrettyWithOpts(stmts[0], 40, false, 2)
```
Comment (knz) [inline pkg/server/statements.go]: As I write in my comment this call should not be made here and instead exist in a separate RPC endpoint.

Also you are using an obsolete version. You should use the new version (with a rebase) and pass the argument `simplify = true` to get rid of the extraneous parentheses.

**Comment by couchand** [comment]
> we should have a RPC endpoint to pretty-print stuff, and the UI code should call that endpoint lazily to render things on demand.

That's not crazy in theory, but consider that we probably need a pretty-printed version of each statement anyway.  It would be horrible UX to have the tooltip wait on a network round-trip before showing.  Even that's assuming we have time to improve the summary so we don't need to put the actual text of the query in the row, in which case we'd definitely need every statement printed.

I guess we could only pretty-print on the details page, not on the list page, but that just seems sad.

> knz commented 7 hours ago
> That's unreasonable.
> 
> mjibson commented 7 hours ago
> Yes, agreed.

Suggestions welcome on the right way to have a pretty form of each statement in the list, unless the very ask is "unreasonable"...

**Comment by knz** [comment]
> Suggestions welcome on the right way to have a pretty form of each statement in the list, unless the very ask is "unreasonable"...

Paginate your display. Only pretty-print per page.

**Comment by knz** [comment]
Another aspect in this conversation: Matt and I have been iterating on the best ways to lay out the SQL code. We have not reached consensus yet. I am not keen on exposing a particular layout to users just yet until we are satisfied with the results.

**Pull request closed without merging**